File: core-di/src/main/java/home/oleg/placesnearme/AppApiProvider.java
package home.oleg.placesnearme;

import android.app.Activity;
import android.app.Application;
import android.support.v4.app.Fragment;

import home.oleg.placesnearme.api.AppApi;

public interface AppApiProvider {
    AppApi getAppApi();

    class Initializer {
        public static AppApi getAppApi(Activity activity) {
            Application application = activity.getApplication();
            return getAppApiInternal(application);
        }

        public static AppApi getAppApi(Fragment fragment) {
            Application application = fragment.getActivity().getApplication();
            return getAppApiInternal(application);
        }

        private static AppApi getAppApiInternal(Application application) {
            if (application instanceof AppApiProvider) {
                return ((AppApiProvider) application).getAppApi();
            } else {
                throw new IllegalStateException("application is not an instance of AppApiProvider");
            }
        }
    }
}

File: core-di/src/main/java/home/oleg/placesnearme/coredi/AppApiProvider.kt
package home.oleg.placesnearme.coredi

import android.app.Activity
import android.app.Application
import androidx.fragment.app.Fragment
import home.oleg.placesnearme.coredi.api.AppApi

interface AppApiProvider {
    val appApi: AppApi

    object Initializer {
        
        fun getAppApi(activity: Activity): AppApi {
            val application = activity.application
            return getAppApiInternal(application)
        }

        fun getAppApi(fragment: Fragment): AppApi {
            val application = fragment.activity!!.application
            return getAppApiInternal(application)
        }

        private fun getAppApiInternal(application: Application): AppApi {
            return if (application is AppApiProvider) {
                (application as AppApiProvider).appApi
            } else {
                throw IllegalStateException("application is not an instance of AppApiProvider")
            }
        }
    }
}
